compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AnvilMergeAnnotationDetectorCheck.kt

```diff
@@ -14,32 +22,62 @@ import org.jetbrains.kotlin.descriptors.ModuleDescriptor
 import org.jetbrains.kotlin.psi.KtFile
 import java.io.File
 
-@AutoService(CodeGenerator::class)
-internal class AnvilMergeAnnotationDetectorCheck : PrivateCodeGenerator() {
-
-  override fun isApplicable(context: AnvilContext): Boolean = context.disableComponentMerging
-
-  override fun generateCodePrivate(
-    codeGenDir: File,
-    module: ModuleDescriptor,
-    projectFiles: Collection<KtFile>
-  ) {
-    val clazz = projectFiles
-      .classAndInnerClassReferences(module)
-      .firstOrNull {
-        it.isAnnotatedWith(mergeComponentFqName) ||
-          it.isAnnotatedWith(mergeSubcomponentFqName) ||
-          it.isAnnotatedWith(mergeInterfacesFqName) ||
-          it.isAnnotatedWith(mergeModulesFqName)
-      }
+internal object AnvilMergeAnnotationDetectorCheck : AnvilApplicabilityChecker {
+  private const val MESSAGE = "This Gradle module is configured to ONLY generate code with " +
+    "the `disableComponentMerging` flag. However, this module contains code that " +
+    "uses Anvil @Merge* annotations. That's not supported."
+  private val ANNOTATIONS_TO_CHECK = setOf(
+    mergeComponentFqName,
+    mergeSubcomponentFqName,
+    mergeInterfacesFqName,
+    mergeModulesFqName
+  )
+  override fun isApplicable(context: AnvilContext) = context.disableComponentMerging
+
+  internal class KspGenerator(
+    override val env: SymbolProcessorEnvironment,
+  ) : AnvilSymbolProcessor() {
+    @AutoService(SymbolProcessorProvider::class)
+    class Provider : AnvilSymbolProcessorProvider(AnvilMergeAnnotationDetectorCheck, ::KspGenerator)
+
+    override fun processChecked(resolver: Resolver): List<KSAnnotated> {
+      val clazz = ANNOTATIONS_TO_CHECK
```

Comment:
Let's make this a sequence so it lazily resolves the types below

```suggestion
      val clazz = ANNOTATIONS_TO_CHECK
        .asSequence()
```

Reply:
I've tested the change locally and pushed the change as suggested. 🙇